scripts/generate-compatibility-json.py

```diff
@@ -0,0 +1,329 @@
+
+    If no custom version is specified, the test suite level version is used to determine if the test should run.
+    """
+
+    specified_versions_str = file_fields.get(f"{test_name}:{FROM_VERSIONS}")
+    test_semver_version = parse_version(version)
+
+    # no custom version defined for this test, run it as normal using the from_version specified on the test suite.
+    if specified_versions_str is None:
+        # if there is nothing specified for this particular test, we just compare it to the version
+        # specified at the test suite level.
+        test_suite_level_version = file_fields[FROM_VERSION]
+        return test_semver_version >= parse_version(test_suite_level_version)
+
+    specified_versions = specified_versions_str.split(",")
+
+    for v in specified_versions:
+        semver_v = parse_version(v)
+        # if the major and minor versions match, there was a specified release line for this version.
+        # do a comparison on that version to determine if the test should run.
+        if semver_v.major == test_semver_version.major and semver_v.minor == test_semver_version.minor:
+            return semver_v >= test_semver_version
+
+    # there was no version defined for this version's release line, but there were versions specified for other release
+    # lines, we assume we should not be running the test.
+    return False
+
+
+def _get_ibc_go_releases(from_version: str) -> List[str]:
+    releases = []
+
+    from_version_semver = parse_version(from_version)
+
+    # ref: documentation https://docs.github.com/en/rest/releases/releases?apiVersion=2022-11-28#list-releases
+    resp = requests.get(RELEASES_URL, params={"per_page": 1000})
+    resp.raise_for_status()
+
+    response_body = resp.json()
+
+    all_tags = [release["tag_name"] for release in response_body]
+    for tag in all_tags:
+        # skip alphas, betas and rcs
+        if any(t in tag for t in ("beta", "rc", "alpha", "icq")):
+            continue
+        try:
+            semver_tag = parse_version(tag)
+        except ValueError:  # skip any non semver tags.
+            continue
+
+        # get all versions
+        if semver_tag <= from_version_semver:
+            releases.append(tag)
+
+    return releases
+
+
+def _build_file_metadata(file_name: str) -> Dict:
+    """_build_file_metadata constructs a dictionary of metadata from the test file."""
+    file_lines = _load_file_lines(file_name)
+    return {
+        TEST_SUITE: _extract_test_suite_function(file_lines),
+        TESTS: _extract_all_test_functions(file_lines),
+        FIELDS: _extract_script_fields(file_lines)
+    }
+
+
+def _extract_test_suite_function(file_lines: List[str]) -> str:
+    """extracts the name of the test suite function in the file. It is assumed
+    there is exactly one test suite defined"""
+    for line in file_lines:
+        line = line.strip()
+        if "(t *testing.T)" in line:
+            return re.search(r"func\s+(.*)\(", line).group(1)
```

Comment:
a little bit gross, we can try and come up with a cleaner way of doing this if people feel strongly about it.

Reply:
if it ain't broke